### PR TITLE
soc: riscv: Add ability for SOC to use custom sys_io functions

### PR DESCRIPTION
**Pull request by ycsin**
Add Kconfig `SOC_HAS_CUSTOM_SYS_IO` symbol that an SoC can set to specify that it has a custom implementation for sys_io functions.

### CODE REVIEW

**Review by carlocaione — CHANGES_REQUESTED** [review]
Why do you need this?

**Comment by ycsin** [comment]
> we usually prefer to have some kind of `z_soc_*` overriding functions

Thanks for the hint, I'll try to get this done properly

**Comment by fkokosinski** [comment]
@ycsin You could check out how LiteX VexRiscv handles it, as it has custom sys_io-style functions implementations: https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/riscv/litex-vexriscv/soc.h

**Comment by carlocaione** [comment]
> @ycsin You could check out how LiteX VexRiscv handles it, as it has custom sys_io-style functions implementations: https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/riscv/litex-vexriscv/soc.h

yeah no, please don't. Introducing some kind of SoC-specific IO helpers is making all the drivers basically un-portable on any different platform. Better sticking to `sys_io` functions and override those when needed.

**Comment by gmarull** [comment]
> @ycsin You could check out how LiteX VexRiscv handles it, as it has custom sys_io-style functions implementations: https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/riscv/litex-vexriscv/soc.h

I'd say this model should be avoided. If we have SoCs that require custom sys_io functions, we should modify sys_io so that it allows for both, arch-specific and/or soc-specific implementations. sys_io.h header chain is a bit broken, so it may require some prep-work.

**Comment by ycsin** [comment]
I've updated the PR to look something like https://github.com/zephyrproject-rtos/zephyr/commit/43a61329cc454ab318eefffc35f78487ff680be2

> we should modify sys_io so that it allows for both, arch-specific and/or soc-specific implementations. sys_io.h header chain is a bit broken, so it may require some prep-work.

you mean we will probably first need some code reorg so that `sys_write8` calls `arch_sys_write8` which then calls `soc_sys_write8` in the end, and allow arch/soc -specific custom implementation to provide `_z_soc_sys_*` & `_z_arch_sys_*`?

File: include/zephyr/arch/common/sys_io.h [hunk 1]
```diff
@@ -15,39 +15,66 @@
 #include <zephyr/toolchain.h>
 #include <zephyr/types.h>
 #include <zephyr/sys/sys_io.h>
+#ifdef CONFIG_SOC_HAS_CUSTOM_SYS_IO
+#include <soc.h>
+#endif
```
Comment (carlocaione) [inline include/zephyr/arch/common/sys_io.h:20]: What about instead:
1. Making this only for RISC-V, other architectures really do not need this
2. Adding a `sys_io.h` file or similar somewhere in `include/zephyr/arch/riscv/`
3. Coping `sys_io.h` from this PR into the RISC-V specific `sys_io.h`
4. In that file adding something like:
```c
#ifdef CONFIG_SOC_HAS_CUSTOM_SYS_IO
extern uint8_t z_soc_sys_read8 ...
...
#endif
```
5. Maybe calling the symbol something like `CONFIG_RISCV_SOC_HAS_..`

So you don't need to have a `soc.h` file just for this and your SoC can only carrying a `.c` file implementing the functions that should be linked in place at compile time.

**Review by carlocaione — CHANGES_REQUESTED** [review]

**Review by carlocaione — CHANGES_REQUESTED** [review]
What about 64-bit operations?

**Comment by ycsin** [comment]
> What about 64-bit operations?

Thanks for the reminder, just added & rebased

**Review by carlocaione — APPROVED** [review]

**Review by fkokosinski — APPROVED** [review]

**Pull request merged**